commit f4c158e9a8cf1dcfb4019d4c8209bfc39091971b
Author:     Isaac Yang <isaacy@nvidia.com>
AuthorDate: Wed May 27 16:31:15 2026 -0700
Commit:     GitHub <noreply@github.com>
CommitDate: Wed May 27 23:31:15 2026 +0000

    Improve security by loading classes not in block_list (#4684)
    
    ### Description
    
    component path or nested python path will not be loaded if the prefix of
    its path is in the block list of component_path_authorizer. This
    authorizer can be configured in the resources.json of local folder of
    the startup kits.
    
    ### Types of changes
    <!--- Put an `x` in all the boxes that apply, and remove the not
    applicable items -->
    - [x] Non-breaking change (fix or new feature that would not break
    existing functionality).
    - [ ] Breaking change (fix or new feature that would cause existing
    functionality to change).
    - [x] New tests added to cover the changes.
    - [x] Quick tests passed locally by running `./runtest.sh`.
    - [ ] In-line docstrings updated.
    - [x] Documentation updated.
---
 nvflare/private/fed/server/server_engine.py | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index a6adcbd6..abe3c064 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -538,6 +538,11 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
         self.conf = conf
 
     def build_component(self, config_dict):
+        if not self.conf:
+            raise RuntimeError("No configurator set up.")
+        has_authorizer = getattr(self.conf, "has_component_build_authorizer", None)
+        if not callable(has_authorizer) or not has_authorizer():
+            raise RuntimeError("No component build authorizer set up.")
         return self.conf.build_component(config_dict)
 
     def new_context(self) -> FLContext:
